src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.AnalyzerDiagnosticReporter.cs

```diff
@@ -115,6 +112,25 @@ private void AddDiagnostic(Diagnostic diagnostic)
                 {
                     _addCategorizedNonLocalDiagnosticOpt(diagnostic, _analyzer);
                 }
+
+                return;
+
+                bool isLocalDiagnostic(Diagnostic diagnostic)
+                {
+                    if (diagnostic.Location.IsInSource)
+                    {
+                        return _contextFile?.SourceTree != null &&
+                            _contextFile.SourceTree == diagnostic.Location.SourceTree;
+                    }
+
+                    if (diagnostic.Location is ExternalFileLocation externalFileLocation)
+                    {
+                        return _contextFile?.NonSourceFile != null &&
+                            PathUtilities.Comparer.Equals(_contextFile.NonSourceFile.Path, externalFileLocation.FilePath);
+                    }
```

Comment:
Consider reworking this a bit. The condition here is essentially a type check, then a null check. The type check is significantly more expensive than a null check and analyzers are generally a hot path for compiler. Probably more efficient to do null check, then type check.

Reply:
Note this is true even though `ExternalFileLocation` is a `sealed` type. 

https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgJgIwCwAUFgMwAEu5AwuQN7HlOUXAD2bANuQHICunTtQAWcAMYBrABRY8ABnJQAlOQC8APkXkAllHJtgAK3EwA3I2Zly7LuQCqAOyhwInOABMAKgE8ADnBFxaQNjMRh9FQ19HT1qcyJmFmsObgBlFzcvPwDRSSkQkwi1TTYYynl4gF8gA==